File: src/algorithms/python/Count_Complete_Tree_Nodes.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def get_height(self, path):
        node, h, tmp = self.root, 0, 1 << (self.max_h - 2)
        while node:
            if tmp & path:
                node = node.right
            else:
                node = node.left
            tmp >>= 1
            h += 1
        return h

    # @param {TreeNode} root
    # @return {integer}
    def countNodes(self, root):
        if root is None:
            return 0
        if root.left is None:
            return 1
        self.root, self.max_h = root, 0
        node = root
        while node:
            node = node.left
            self.max_h += 1
        l, r = 0, (1 << (self.max_h - 1)) - 1
        while l < r - 1:
            mid = (l + r) >> 1
            _h = self.get_height(mid)
            if _h == self.max_h:
                l = mid
            else:
                r = mid
        l_h, r_h = self.get_height(l), self.get_height(r)
        if l_h == r_h:
            return 2 ** self.max_h - 1
        elif l_h == self.max_h - 1:
            return 2 ** (self.max_h - 1) + l - 1
        else:
            return 2 ** (self.max_h - 1) + r - 1


if __name__ == "__main__":
    solution = Solution()
    l = [1, 2, 3, 4]
    node = map(TreeNode, l)
    node[0].left, node[0].right = node[1], node[2]
    node[1].left = node[3]
    print solution.countNodes(node[0])

File: src/algorithms/cpp/Count_Complete_Tree_Nodes.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    int len;
    
private:
    int get_high(TreeNode *root, int m) {
        TreeNode *node = root;
        int tmp_len = 1, tmp = (1 << (len - 2));
        for (int i = 0; i < len - 1; ++ i) {
            if ((m & tmp) > 0)   node = node->right;
            else                 node = node->left;
            if (node == NULL)   break;
            ++ tmp_len;
            tmp >>= 1;
        }
        return tmp_len;
    }

public:
    int countNodes(TreeNode* root) {
        if (root == NULL)       return 0;
        len = 0;
        TreeNode *node = root;
        while (node) {
            node = node->left;
            ++ len;
        }
        int l = 0, r = (1 << (len - 1)) - 1;
        cout << l << " " << r << endl;
        while (l < r - 1) {
            int m = (l + r) >> 1, tmp_len = get_high(root, m);
            if (tmp_len < len)  r = m;
            else                l = m;
        }
        int r_high = get_high(root, r), l_high = get_high(root, l);
        if (r_high == l_high)   return pow(2, len) - 1;
        else                    return pow(2, len - 1) + l;
    }
};

int main() {
    return 0;
}
